sklearn/metrics/cluster/_supervised.py

```diff
@@ -156,29 +158,67 @@ def contingency_matrix(
     """
 
+    xp, _, device_ = get_namespace_and_device(labels_true, labels_pred)
+
     if eps is not None and sparse:
         raise ValueError("Cannot set 'eps' when sparse=True")
 
-    classes, class_idx = np.unique(labels_true, return_inverse=True)
-    clusters, cluster_idx = np.unique(labels_pred, return_inverse=True)
+    classes, class_idx = xp.unique_inverse(labels_true)
+    clusters, cluster_idx = xp.unique_inverse(labels_pred)
+    c_x = class_idx.shape[0]
     n_classes = classes.shape[0]
     n_clusters = clusters.shape[0]
-    # Using coo_matrix to accelerate simple histogram calculation,
-    # i.e. bins are consecutive integers
-    # Currently, coo_matrix is faster than histogram2d for simple cases
-    contingency = sp.coo_matrix(
-        (np.ones(class_idx.shape[0]), (class_idx, cluster_idx)),
-        shape=(n_classes, n_clusters),
-        dtype=dtype,
-    )
+
+    if _is_numpy_namespace(xp):
+        # Using coo_matrix to accelerate simple histogram calculation,
+        # i.e. bins are consecutive integers
+        # Currently, coo_matrix is faster than histogram2d for simple cases
+        contingency = sp.coo_matrix(
+            (np.ones(c_x), (class_idx, cluster_idx)),
+            shape=(n_classes, n_clusters),
+            dtype=dtype,
+        )
+        if sparse:
+            contingency = contingency.tocsr()
+            contingency.sum_duplicates()
+        else:
+            contingency = contingency.toarray()
+            if eps is not None:
+                # don't use += as contingency is integer
+                contingency = contingency + eps
+        return contingency
+
+    # Array API support
     if sparse:
-        contingency = contingency.tocsr()
-        contingency.sum_duplicates()
-    else:
-        contingency = contingency.toarray()
-        if eps is not None:
-            # don't use += as contingency is integer
-            contingency = contingency + eps
+        raise ValueError(
+            f"Array API dispatch to namespace {xp.__name__} only supports "
+            f"sparse=False for contingency_matrix."
+        )
+
+    floating_dtype = _max_precision_float_dtype(xp, device_)
+    contingency = xp.zeros(
+        (n_classes, n_clusters), device=device_, dtype=floating_dtype
+    )
+    arange_classes = xp.arange(n_classes, device=device_, dtype=class_idx.dtype)
+    arange_clusters = xp.arange(n_clusters, device=device_, dtype=cluster_idx.dtype)
+
+    batch_size = 1024
+    for i in range(0, c_x, batch_size):
+        i_end = min(i + batch_size, c_x)
+        batch_classes = class_idx[i:i_end, ...]
+        batch_clusters = cluster_idx[i:i_end, ...]
+
+        batch_classes_one_hot = xp.astype(
+            batch_classes[:, None] == arange_classes, floating_dtype
+        )
+        batch_clusters_one_hot = xp.astype(
+            batch_clusters[:, None] == arange_clusters, floating_dtype
+        )
+
+        contingency += batch_classes_one_hot.T @ batch_clusters_one_hot
```

Comment:
Alternatively, we could use `sklearn.utils._array_api._bincount` to do this instead:

```python
        weights = xp.ones(c_x, dtype=count_dtype, device=device_)
        flat_indices = class_idx * n_clusters + cluster_idx
        contingency_flat = _bincount(
            flat_indices,
            weights=weights,
            minlength=n_classes * n_clusters,
            xp=xp,
        )
        contingency = xp.reshape(contingency_flat, (n_classes, n_clusters))
```

However, [_bincount](https://github.com/scikit-learn/scikit-learn/blob/14162d45062e84a4c93ccbef9edb885585886830/sklearn/utils/_array_api.py#L1031-L1037) is just converting to a numpy array and using `np.bincount` for namespaces without a `bincount` implementation.